### PR TITLE
[WOR-433] Handle custom date and time partition column names on Broad spend report table.

### PR DESCRIPTION
**Pull request by cahrens**
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala [hunk 1]
```diff
@@ -308,6 +308,7 @@ object Boot extends IOApp with LazyLogging {
       val submissionCostService: SubmissionCostService =
         SubmissionCostService.constructor(
           gcsConfig.getString("billingExportTableName"),
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala:310]: These values are added to rawls.conf in https://github.com/broadinstitute/firecloud-develop/pull/3051

File: core/src/main/scala/org/broadinstitute/dsde/rawls/model/SpendReportingModel.scala [hunk 2]
```diff
@@ -52,7 +52,7 @@ object SpendReportingAggregationKeys {
   }
 
   case object Daily extends SpendReportingAggregationKey {
-    override val bigQueryField: String = "DATE(_PARTITIONTIME)"
+    override val bigQueryField: String = "DATE(REPLACE_TIME_PARTITION_COLUMN)"
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/model/SpendReportingModel.scala:55]: I don't love this, but I tried pushing the configuration value down into this class and it was an absolute mess. This approach allows us to keep the logic of "use default config value or not" in a single place-- https://github.com/broadinstitute/rawls/pull/1963/files#diff-6bdd7e72eedead7aafdcd9b71fc64d8eaf45c390b2db2d004f768ea07a43fc53R220

File: core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala [hunk 3]
```diff
@@ -217,18 +217,21 @@ class SpendReportingService(userInfo: UserInfo, dataSource: SlickDataSource, big
       .build()
   }
 
-  private def getQuery(aggregationKeys: Set[SpendReportingAggregationKey], tableName: String): String = {
-    s"""
+  def getQuery(aggregationKeys: Set[SpendReportingAggregationKey], tableName: String, customTimePartitionColumn: Option[String]): String = {
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:220]: Exposed this method for unit testing.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala [hunk 4]
```diff
@@ -217,18 +217,21 @@ class SpendReportingService(userInfo: UserInfo, dataSource: SlickDataSource, big
...
+    // The Broad table uses a view with a different column name.
+    val timePartitionColumn = customTimePartitionColumn.getOrElse("_PARTITIONTIME")
+    val queryClause = s"""
        | SELECT
        |  SUM(cost) as cost,
        |  SUM(IFNULL((SELECT SUM(c.amount) FROM UNNEST(credits) c), 0)) as credits,
        |  currency ${aggregationKeys.map(_.bigQueryAliasClause()).mkString}
-       | FROM `$tableName`
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:226]: I don't know where back ticks were used-- any reason @marctalbott?

**Comment by cahrens** [comment]
Swatomation test failures are related to https://github.com/broadinstitute/firecloud-develop/pull/3051 not being merged yet.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala (hunk 4 above)
Comment (marctalbott) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:226]: Does it work without back ticks? Seems to be pretty standard to wrap the table name in back ticks both in our code and in [Google's documentation](https://cloud.google.com/bigquery/docs/running-queries#queries) (see Java section of example)

Comment (cahrens) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:226]: Yes, it works without back ticks, but I will add it back in.

**Review by aherbst-broad — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Comment by cahrens** [comment]
jenkins retest

**Pull request merged**